internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,163 @@ import (
 	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
 )
 
+
+
+// Runtime2 is the VM.
+// Review: the object that implements this outlives actor calls.
+type Runtime2 struct {
+	CurerntEpoch() Epoch
+	// Review: why is this taking an offset? verify size of offset
+	Randomness(epoch Epoch, offset uint64) Randomness
+	// Review: the spec still has a TODO on this
+	// Review: not sure why the original has a messagereceipt instead of an InvocOutput
+	Send(input InvocInput) InvocOutput
+	// Review: how does this work? What qualifies as an "error"?
+	// Review: how is this any different than the caller just ignoring a returned error?
+	// SendAllowingErrors(input msg.InvocInput) msg.MessageReceipt
+
+	// Review: we might want a global Storage() Storage here
+}
+
+// InvocationContext is passed to the actors on each method call.
+// Review: this type + Runtime2 is mostly what the spec defines as Runtime.
+type InvocationContext interface {
+	// Review: grouped the VM level things out of the invocation itself
+	Runtime() Runtime2
+	Caller() address.Address
+	// Review: (is in spec, dont think we need it) idem to Caller().Equal(addr) :/
+	// ValidateCallerIs(caller addr.Address)
```

Comment:
The idea here is that the runtime can _enforce_ that the actor code calls `ValidateCallerIs` (or `ValidateCallerMatches`), and automatically raise an error if it's never called. (The rationale being that by default, methods are callable by any party introducing a message to the chain, and there is a large class of security-critical bugs that could arise if actors, particularly system actors, don't check the identity of the caller.)

Reply:
Not sure this gets you want you want.. you are just enforcing that a call existed, not that it is the correct call. It sill requires the same security review. 
What are you doing about the methods that do NOT have a requirement? How in the runtime telling them apart? 
How will the runtime know you are validating the correct address?

From a security review standpoint, reviewing..
```
runtime.Caller().IsActor(actor.FooBar)
// super sensitive code
```

..or reviewing..
```
runtime.ValidateCallerIs(actor.FooBar)
// super sensitive code
```

..is the same.

If the spec requires certain methods to be only callable by a specific `Actor`, it might be best for it to just say that, and indicate which methods fall under this restriction. 

Implementations can decide how they want to deal with that (interfaces, method annotations, in-line checks, etc.).

Reply:
- It's an additional safety check. It has exactly the same review cost, but just in case security review misses some control flow path that completely fails to check the caller identity, we would still prefer to "fail safe" than allow the method to complete execution.

- For methods that don't need to check the caller identity, they should call `ValidateCallerMatches(...accept-everything pattern...)` to make that clear in a way that's (to the extent possible) checkable by the runtime.

Reply:
How would a reviewer fail to read that line? shouldn't be a check at the top of the method? 
Or is the spec intending to have dynamic checks on the code depending on input arguments?

Reply:
Yes, ideally it should always be a simple check at the top of the method (and its absence should easily be caught during security review), but we'd like to allow for the possibility of more complex checks in the future. (And even if we don't ever need more complex checks, I think it still doesn't hurt to have this extra check in the runtime, just in case.)